### ISSUE TITLE
How to accept the server returns the response?

### ISSUE THREAD

**Issue by LuisKen** [bleenco/ngx-uploader#421]

**Comment by jkuri** [comment]
can you clarify the question?

**Comment by LuisKen** [comment]
ok，after the files uploaded to the background，Server side forward pass a result, the result in the response,
But I don't know how to use this plug-in to receive the response。
Like this：This is sends a POST request to the background，But I don't know how to receive the background the returned result。

startUpload(): void {
    const event: UploadInput = {
      type: 'uploadAll',
        url: 'localhost:8080/servie/upload',
        method: 'POST',
      data: { foo: 'bar' }
    };

    this.uploadInput.emit(event);
  }

**Comment by jkuri** [comment]
hi. you need to "catch" events that are emitted like its done [here](https://github.com/bleenco/ngx-uploader/blob/master/src/components/app-home/app-home.component.ts#L23). when you got response from server it should have `output.type === 'done'`, just `console.log(output);` to see which values are available and you will know how to proceed.

**Comment by LuisKen** [comment]
Well, Thank you for your answer, but the problem hasn't solved，add this code to     onUploadOutput(output: UploadOutput): 
else if (output.type === 'done') {
  this.files.forEach(file => {
    console.log(file.response);
  });
}
But in onUploadOutput this method, To track the upload process，“output.type”’ respectively：addedToQueue、allAddedToQueue、start、uploading，then this method is over，how should I do？

**Comment by LuisKen** [comment]
When the upload is completed，no longer automatically invoke onUploadOutput ，that is “else if (output.type === 'done')” never be executed，what reason is this？I didn't see similar problems on stack overflow   ，this is my code：

onUploadOutput(output: UploadOutput): void {
    if (output.type === 'allAddedToQueue') {
        ....
    }else if(...){...
    }else if (output.type === 'done') {
        ....
    }
}

**Comment by jkuri** [comment]
I am not sure I understand your question, did you checked README file?

**Comment by LuisKen** [comment]
Yes, I read it,and all the upload process of mine is successful,and has been through a browser to see the server returned results.Hm,I mean, In my "onUploadOutput" method,from beginning to end，“output.type”  doesn't have ‘done’ state

**Comment by jkuri** [comment]
are you sure your server returns the the response?

**Comment by LuisKen** [comment]
Yes, I sent an email to you, there is a screenshot about this problem, you received?

**Comment by jkuri** [comment]
I didn't. You can attach screenshot here, just drag&drop into textarea field.

**Comment by jkuri** [comment]
by the way... are you using pace.js in your application? some users reported that it doesn't work in a combination with ngx-uploader module.

**Comment by LuisKen** [comment]
![Uploading 3.png…]()
![Uploading 2.png…]()
![Uploading 1.png…]()
![Uploading code.png…]()

**Comment by LuisKen** [comment]
![4](https://user-images.githubusercontent.com/20528035/35611611-92ded58c-06a0-11e8-9461-af04aecbaaf1.png)
![3](https://user-images.githubusercontent.com/20528035/35611613-93132cf6-06a0-11e8-8037-63696435600d.png)
![2](https://user-images.githubusercontent.com/20528035/35611614-934a4434-06a0-11e8-897e-679507f728c6.png)
![1](https://user-images.githubusercontent.com/20528035/35611615-93848f9a-06a0-11e8-8165-6d1d76cba9e1.png)
![code](https://user-images.githubusercontent.com/20528035/35611616-93be21ce-06a0-11e8-91ab-e3cfcce9addf.png)

**Comment by LuisKen** [comment]
Hi，my problems are similar to this：
![111](https://user-images.githubusercontent.com/20528035/35612686-9e458318-06a4-11e8-9542-8bd19d341165.png)

**Comment by jkuri** [comment]
you need to `console.log(output)` and not `this.files`.

**Comment by LuisKen** [comment]
Yes, I am the output, or there is no "done" status
![88888](https://user-images.githubusercontent.com/20528035/35654720-3ac216fa-0729-11e8-9276-2679e6410b79.png)

**Comment by jorgelub12** [comment]
The problem is that event done is never triggered. Any fix on this ?

**Comment by sgyy** [comment]
https://github.com/bleenco/ngx-uploader/blob/master/src/ngx-uploader/classes/ngx-uploader.class.ts#L187
use   number  4  replace  XMLHttpRequest.DONE 
if (xhr.readyState === 4)

**Comment by MatzeSchmitt** [comment]
@sgyy Can you please specify your answer? What to do and where? 

Also facing this issue right know. Is there a solution in the meantime?

Thanks.

**Comment by Matze-Schmitt** [comment]
Facing the same issue here.

Can someone please explain how to properly catch the server response after uploading?

@sgyy I don´t really want to change any code from inside the node_modules.

**Comment by dpollastrini** [comment]
@Matze-Schmitt Unfortunately, the source of this bug is likely https://github.com/HubSpot/pace/issues/261 and manifests in ngx-uploader as describe in https://github.com/bleenco/ngx-uploader/issues/408.  Bottom line is pace needs to be fixed, not necessarily ngx-uploader (which is not the only component that's affected).  Pace is breaking a pretty basic component in their override of xmlhttprequest.

**Comment by mariusl-a** [comment]
**I don't use pace for my application, this worked for me:** 

In your onUploadOutput function: Add `else if(output.type === 'done') { //output.file.response }`
Added my entire onUploadOutput function incase of any confusion.
Good luck =)

```
onUploadOutput(output: UploadOutput): void {
    if (output.type === 'allAddedToQueue') {
      let token = this.authService.getToken(); 

      const event: UploadInput = {
        type: 'uploadAll',
        url: this.config.apiEndpoint + '/api/v1/attachments/upload',
        method: 'POST',
        headers: { 'Authorization': token }
      };
      this.uploadInput.emit(event);

    } else if (output.type === 'addedToQueue' && typeof output.file !== 'undefined') { // add file to array when added
      this.files.push(output.file);
    } else if (output.type === 'uploading' && typeof output.file !== 'undefined') {
      this.uploading = true;
      // update current data in files array for uploading file
      const index = this.files.findIndex(file => typeof output.file !== 'undefined' && file.id === output.file.id);
      this.files[index] = output.file;
    } else if (output.type === 'removed') {
      // remove file from array when removed
      this.files = this.files.filter((file: UploadFile) => file !== output.file);
    } else if (output.type === 'dragOver') {
      this.dragOver = true;
    } else if (output.type === 'dragOut') {
      this.dragOver = false;
    } else if (output.type === 'drop') {
      this.dragOver = false;
    } else if(output.type === 'done') {
      this.addFile(output.file.response);
    } else if(output.type === 'rejected') {
      console.log('Rejected', output);
    }
  }
```

- Angular 6.1.7 
- ngx-uploader 6.0.2

**Comment by Gullfaxi171** [comment]
I had the same issue, fact is that the output variable is bound to the <input type="file">

If you hide the input with *ngIf right after starting the upload, you also destroy the variable output, and will never catch the "done" event
